packages/picasso/src/utils/use-body-scroll-lock.ts

```diff
@@ -0,0 +1,53 @@
+import { isBrowser } from '@toptal/picasso-shared'
+import { useEffect, useMemo } from 'react'
+
+const layers = new Set<number>();
+let scrollLock: { prevBodyOverflow: string } | undefined = undefined;
+
+export const useBodyScrollLock = (isLocked: boolean) => {
+  const layerId = useMemo(generateLayerId, [])
+
+  useEffect(() => {
+    if (isLocked) {
+      layers.add(layerId)
+      dropScrollLock()
+    }
```

Comment:
Shouldn't we also remove the layer in the case of `!isLocked`, imagine that the variable would flip mid-execution by some conditional we would need to remove the layer. (even though it should be rare, we need to think on it). We would also need to check if all the layers are removed and accordingly remove lift the scroll-lock, something like:
```suggestion
    if (isLocked) {
      layers.add(lockedLayerId)
    } else {
      layers.remove(lockedLayerId)
    }
    syncScrollLock()
```

Where `checkForScrollLock()` would be:
```js
const syncScrollLock = () => {
  if (layers.size > 0) {
    doScrollLock()
  } else {
    liftScrollLock()
  }
}
```

Reply:
>Shouldn't we also remove the layer in the case of !isLocked, imagine that the variable would flip mid-execution...

this is totally covered by how useEffect works. any change to isLocked will result in useEffect being unmounted and mounted again. During unmount it will run cleanup callback (2) - which removes component instance from layers and syncs scroll lock, then on next mount with new isLocked=false component instance isn't added to layers thanks to condition (1). 

![image](https://user-images.githubusercontent.com/903437/212138908-aa9d19fe-08d5-46f3-a1a8-b1e5fcd038f8.png)

Reply:
You are right, I got confused with the global state stuff, the cleanup callback should do the work! Thanks @ascrazy! :+1: